posts/2021-08-31-liberty-bikes.adoc

```diff
@@ -0,0 +1,76 @@
+
+[#Implementation]
+== How It Works
+
+image::img/blog/lb-block-diagram.png[block diagram of Liberty Bikes, showing how the microservices communicate, align="center"]
+
+link:https://github.com/OpenLiberty/liberty-bikes[Liberty Bikes is implemented] as three backend microservices that power the link:https://github.com/OpenLiberty/liberty-bikes/tree/master/game-service[game itself] and the link:https://github.com/OpenLiberty/liberty-bikes/tree/master/player-service[player management and leaderboard features], as well as link:https://github.com/OpenLiberty/liberty-bikes/tree/master/auth-service[authenticating users]. Each of these services is a MicroProfile and Java-based application running on Open Liberty. The game's user interface is an Angular-based https://github.com/OpenLiberty/liberty-bikes/tree/master/frontend[front end application].
+
+Data flow among the services is best described by walking through what happens when you access the game. 
+
+image::img/blog/lb-login-page-start.png[screen capture of the Liberty Bikes login page, align="center"]
+
+First, when you load the game, the front end user interface makes a REST API call to the Authentication microservice to link:https://github.com/OpenLiberty/liberty-bikes/blob/master/auth-service/src/main/java/org/libertybikes/auth/service/AuthTypes.java[find out] which OAuth authentication providers have been configured (all REST APIs in Liberty Bikes are implemented with JAX-RS and MicroProfile OpenAPI). If you choose to log in with an available provider, the UI makes more REST calls to the Authentication service to set up the flow. 
+
+Once you are authenticated, the UI then makes a REST call to the Player microservice to link:https://github.com/OpenLiberty/liberty-bikes/blob/master/player-service/src/main/java/org/libertybikes/player/service/PlayerService.java[add you] to the Players database. Now you're logged in.
+
+image::img/blog/lb-logged-in-page.png[screen capture of a player logged in to Liberty Bikes ready to start a game, align="center"]
+
+Suppose you click Play Now to quickly get into a game. The UI makes a REST call to the Game microservice to link:https://github.com/OpenLiberty/liberty-bikes/blob/master/game-service/src/main/java/org/libertybikes/game/round/service/GameRoundService.java[create] a new game room, and then sets up the game board by opening a link:https://github.com/OpenLiberty/liberty-bikes/blob/master/game-service/src/main/java/org/libertybikes/game/round/service/GameRoundWebsocket.java[websocket] between your browser and the Game service; this allows quick communication of the location of players and obstacles every frame, as well as sending your control inputs to the Game service so it knows which direction to move you in the next frame. Since we want to update the board every 50 milliseconds, speed is essential.
```

Comment:
Might consider breaking the one long sentence up into a couple shorter ones:

The UI makes a REST call to the Game microservice to create a new game room, and then sets up the game board by opening a websocket between your browser and the Game service; this allows quick communication of the location of players and obstacles every frame, as well as sending your control inputs to the Game service so it knows which direction to move you in the next frame. Since we want to update the board every 50 milliseconds, speed is essential.

--->

The UI makes a REST call to the Game microservice to create a new game room, and then sets up the game board by opening a websocket between your browser and the Game service. This connection sends your control inputs to the Game service so it knows which direction to move you in the next frame. It also allows quick communication of the location of players and obstacles in every frame. Since we want to update the board every 50 milliseconds, speed is essential.